### DISCUSSION TITLE
Why doesn't the `OpenAPI` specification include the response for `raise HTTPException`?

### DISCUSSION THREAD

**Discussion by sagardwivedi** [fastapi/fastapi#11882] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, HTTPException

app = FastAPI()

@router.post("/login/access-token")
def login_access_token(
    session: SessionDep,
    form_data: Annotated[OAuth2PasswordRequestForm, Depends()],
) -> Token:
    """
    Get an access token for authentication.

    This endpoint authenticates a user's credentials and returns an access token
    if the credentials are valid.

    - `session`: SQLModel database session dependency.
    - `form_data`: OAuth2PasswordRequestForm object containing username and password.

    ## Returns:
    - `Token`: A Token object containing the access token and token type.

    ## Raises:
    - `HTTPException`: If the provided username or password is incorrect.
    """
    user = authenticate(
        session=session, username=form_data.username, password=form_data.password
    )
    if not user:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Incorrect email or password",
        )
    access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(user.user_id, access_token_expires)
    return Token(access_token=access_token, token_type="Bearer")
```

### Description

I wrote a login API, and when I enter the wrong password, it sends back a status code 400 along with a message. However, I know that the OpenAPI specification doesn't directly include the `raise HTTPException` feature. Is there a workaround for this?

### Problem Details

- **Current Behavior**: When a user enters the wrong password, the API responds with a status code 400 and a corresponding error message.
- **Issue**: The OpenAPI specification does not inherently support the `raise HTTPException` used in FastAPI for error handling. This means that the error handling as implemented does not directly translate to OpenAPI documentation.

### Desired Outcome

To have the OpenAPI documentation accurately reflect the API's behavior, including the handling of incorrect passwords and the resulting status code 400 response. So that `@hey-api/openapi-ts` correct properly.

### Question

Is there a workaround to properly document this behavior in the OpenAPI specification?

### OpenAPI specification (optional)

``` json
"responses": {
          "200": {
            "description": "Successful Response",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Token"
                }
              }
            }
          },
          "422": {
            "description": "Validation Error",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/HTTPValidationError"
                }
              }
            }
          }
        }
```

### Operating System

Windows

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.12.3

**Comment by YuriiMotov** [comment, marked as answer]
FastAPI only includes 200 and 422 responses automaticaly. You can specify other responses manually as described in this part of the documentation: https://fastapi.tiangolo.com/advanced/additional-responses/

You can also specify common responses for all app's routes or for the group of routes (router).

**Reply by sagardwivedi** [reply]
Thank You. I will try it out.

**Reply by sagardwivedi** [reply]
I try it. It worked but I have to write more code. Can't it do it automatically base the `raise HTTPException`

**Reply by ceb10n** [reply]
To make this work as you want, FastAPI would need to implement sort of a static code analysis before it starts. I don't think it makes sense...

**Comment by haydenseitz** [comment]
Is there any recommendation or best practice on how to define response models based on the HTTPException response?

The HTTPException that originates from Starlette has a top-level property `detail` in the response body, so anything passed to `detail` in an HTTPException is under that prop. My org has code that passes a pydantic model (SQLModel) as the `detail` value of an HTTPException and the route has the same model defined as the response model for the matching error code. We noticed that the docs didn't match the actual response because the model was under the `detail` prop in the exception response.

Example:

```python
# app/models.py
class BookModel(SQLModel):
    id: str
    title: str | None = None
```
```python
# app/main.py
@router.get(
    "/book/{book_id}",
    response_model=BookModel,
    responses={400: {"description": "Error", "model": BookModel}},
)
def get_book(book_id: str):
    if book_id != "test":
        raise HTTPException(
            status_code=400,
            detail=BookModel(id=book_id).dict()),
        )
    return BookModel(id=book_id)
```

In API docs, the HTTP 400 response shows as:
```json
{
  "id": "string",
  "title": "string"
}
```

But if the HTTPException is raised, the response is actually:
```json
{
  "detail": {
    "id": "test1",
    "title": null
  }
}
```

A workaround for this is to create a new pydantic model (SQLmodel) for the HTTPException response:
```python
# app/models.py
class HTTPExceptionResponse(SQLModel):
    detail: Any

class BookModelExceptionResponse(HTTPExceptionResponse):
   detail: BookModel
```
```python
# app/main.py
@router.get(
    "/book/{book_id}",
    response_model=BookModel,
    responses={400: {"description": "Error", "model": BookModelExceptionResponse}},
)
def get_book(book_id: str):
# ...
```

This seems like it would be a common scenario: needing a response model for an HTTPException, even for a basic response like `detail=my error message here`. Is there something in FastAPI or an upstream library that can help with this?

EDIT: There is information [here](https://fastapi.tiangolo.com/advanced/additional-responses/) about using a custom response model like `Message` with `JSONResponse`. I am curious if the same practice should be done when using HTTPException.
